src/pfs/pfs.go

```diff
@@ -341,6 +341,71 @@ func onlyHex(p []byte) bool {
 	return true
 }
 
+func isDigit(b byte) bool {
+	return '0' <= b && b <= '9'
+}
+
+// counts the overall offset of branch root and ancestor of
+// also removes everything of the ancestry reference
+// this function also check following circumstances for ancestry reference:
+// 1. ancestry reference should only contain "^" "." and digits
+// 2. If there is branch root ".", it should be the first operator; It also should be the only "."
+// 3. There should be a number right after "."
+// 4. If branch root and ancestor of are used together, the offset of branch root should >= offset of ancestor of
+func countOffsetAndClean(b *[]byte, offset *uint32) error {
```

Comment:
Some of this validation could be done using a regular expression. However, I think doing it your way is more user-friendly since we can return more specific errors. 👍

Reply:
Great!

Reply:
Actually, I spent a little more time thinking this though and maybe its not worth the extra complexity?

For the record, we could parse it using a regular expression like in the test I wrote below:
```go
func TestRegex(t *testing.T) {
	input := "repo.3^^"
	re := regexp.MustCompile("^([[:alnum:]]+)(@[[:alnum:]]+)?(\\.[[:alnum:]]+)?(\\^*)?$")
	fmt.Println(re.Match([]byte(input)))
	parts := re.FindAllStringSubmatch(input, -1)
	fmt.Println(parts)
	var descendants, ancestors int
	var err error
	for _, part := range parts[0] {
		if len(part) == 0 {
			fmt.Println("skipping part", part)
			continue
		}
		switch part[0] {
		case '^':
			ancestors = len(part)
		case '.':
			descendants, err = strconv.Atoi(part[1:])
			require.NoError(t, err)
		}
		offset := descendants - ancestors
		if offset < 0 {
			t.Fatalf("there should be less ancestors than descendants")
		}
	}
}
```
This uses capture groups to break the expression into parts that we can make assumptions about.

I used this website to write up the regular expression: https://regex101.com/
I fooled around a bit until I came up with: `^([[:alnum:]]+)(@[[:alnum:]]+)?(\.[[:alnum:]]+)?(\^*)?$`
And then I tested against these inputs:
```
alph9a@master
alpha2
alpha@zero@niner
blah@foo.24765757
blah@foo.3^12
blah@foo.3^^
blah@foo.7.8
blah@foo^.9
```
I didn't add code to simplify the expression, but you could do that if part[2] and [3] are not nil. Because there is a static amount of capture groups, the array will always be the same size.

Reply:
EDIT:

Here's a version that simplifies the expression:
```go
func TestRegex(t *testing.T) {
	input := "repo.3^^"
	re := regexp.MustCompile("^([[:alnum:]]+)(@[[:alnum:]]+)?(\\.[[:alnum:]]+)?(\\^*)?$")
	fmt.Println(re.Match([]byte(input)))
	parts := re.FindAllStringSubmatch(input, -1)[0]
	fmt.Println(parts)

	var descendants, ancestors int
	var err error
	baseCommit := parts[1] + parts[2]
	branchRoot := parts[3]
	ancestors = len(parts[4])

	if branchRoot != "" {
		descendants, err = strconv.Atoi(branchRoot[1:])
		require.NoError(t, err)
		offset := descendants - ancestors
		if offset < 0 {
			t.Fatalf("there should be less ancestors than descendants")
		}
		fmt.Printf("%s.%d\n", baseCommit, offset)
	} else {
		fmt.Println(input)
	}
}
```